log/record.go

```diff
@@ -0,0 +1,222 @@
+
+// Record TODO: comment.
+// TODO: Add unit tests.
+type Record struct {
+	timestamp         time.Time
+	observedTimestamp time.Time
+	severity          Severity
+	severityText      string
+	body              string
+
+	// The fields below are for optimizing the implementation of
+	// Attributes and AddAttributes.
+
+	// Allocation optimization: an inline array sized to hold
+	// the majority of log calls (based on examination of open-source
+	// code). It holds the start of the list of attributes.
+	front [attributesInlineCount]attribute.KeyValue
+
+	// The number of attributes in front.
+	nFront int
+
+	// The list of attributes except for those in front.
+	// Invariants:
+	//   - len(back) > 0 iff nFront == len(front)
+	//   - Unused array elements are zero. Used to detect mistakes.
+	back []attribute.KeyValue
+}
+
+const attributesInlineCount = 5
+
+// Severity TODO: comment.
+type Severity int
+
+// TODO: comment.
+const (
+	SeverityUndefined Severity = iota
+	SeverityTrace
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
+	SeverityDebug
+	SeverityDebug2
+	SeverityDebug3
+	SeverityDebug4
+	SeverityInfo
+	SeverityInfo2
+	SeverityInfo3
+	SeverityInfo4
+	SeverityWarn
+	SeverityWarn2
+	SeverityWarn3
+	SeverityWarn4
+	SeverityError
+	SeverityError2
+	SeverityError3
+	SeverityError4
+	SeverityFatal
+	SeverityFatal2
+	SeverityFatal3
+	SeverityFatal4
+)
+
+// Timestamp TODO: comment.
+func (r Record) Timestamp() time.Time {
+	return r.timestamp
+}
+
+// SetTimestamp TODO: comment.
+func (r *Record) SetTimestamp(t time.Time) {
+	r.timestamp = t
+}
+
+// ObservedTimestamp TODO: comment.
+func (r Record) ObservedTimestamp() time.Time {
+	return r.observedTimestamp
+}
+
+// SetObservedTimestamp TODO: comment.
+func (r *Record) SetObservedTimestamp(t time.Time) {
+	r.observedTimestamp = t
+}
+
+// Severity TODO: comment.
+func (r Record) Severity() Severity {
+	return r.severity
+}
+
+// SetSeverity TODO: comment.
+func (r *Record) SetSeverity(s Severity) {
+	r.severity = s
+}
+
+// SeverityText TODO: comment.
+func (r Record) SeverityText() string {
+	return r.severityText
+}
+
+// SetSeverityText TODO: comment.
+func (r *Record) SetSeverityText(s string) {
+	r.severityText = s
+}
+
+// Body TODO: comment.
+func (r Record) Body() string {
+	return r.body
+}
+
+// SetBody TODO: comment.
+func (r *Record) SetBody(s string) {
+	r.body = s
+}
+
+// WalkAttributes calls f on each [attribute.KeyValue] in the [Record].
+// Iteration stops if f returns false.
+func (r Record) WalkAttributes(f func(attribute.KeyValue) bool) {
+	for i := 0; i < r.nFront; i++ {
+		if !f(r.front[i]) {
+			return
+		}
+	}
+	for _, a := range r.back {
+		if !f(a) {
+			return
+		}
+	}
+}
+
+// AddAttributes appends the given [attribute.KeyValue] to the [Record]'s list of [attribute.KeyValue].
+// It omits invalid attributes.
+func (r *Record) AddAttributes(attrs ...attribute.KeyValue) {
```

Comment:
Making a separate thread for https://github.com/open-telemetry/opentelemetry-go/pull/4725#discussion_r1409612729.

The idea is to have `Logger.WithAttributes`  and `Record` would be a simple struct with only exported fields. Then the performance improvements (e.g. to reduce the number of heap allocations) can be implemented by the SDK instead of the API.

The new method would have to be added to the specification.

We first need to double-check if it is possible to create an SDK that would have no heap allocations when emitting record with attributes and compare the benchmark results.

Reply:
I made a porotype of the proposal here: https://github.com/pellared/opentelemetry-go/pull/3#pullrequestreview-1775074173

The main problem I encountered is that when `WithAttributes(attrs ...attribute.KeyValue)` is called (on an interface), Go allocates the passed `attrs` on the heap.

More information in the PR.

Reply:
I made a prototype of the proposal here: https://github.com/pellared/opentelemetry-go/pull/3

The main problem I encountered is that when `WithAttributes(attrs ...attribute.KeyValue)` is called (on an interface), Go allocates the passed `attrs` on the heap.

More information in the PR. I propose to move further discussion to the https://github.com/pellared/opentelemetry-go/pull/3 and here we will just make a final decision.

@MrAlias I am leaning towards rejecting the proposal. However, please double-check the PR if I have not missed anything.

Reply:
TODO: document as a rejected proposal.

Reply:
https://github.com/open-telemetry/opentelemetry-go/pull/4725/commits/9f9acaf260aca5a4677ae71fb80cc29d2a3d78d8